crates/sui/src/client_ptb/builder.rs

```diff
@@ -0,0 +1,993 @@
+            ptb: ProgrammableTransactionBuilder::new(),
+            reader,
+            last_command: None,
+            errors: Vec::new(),
+        }
+    }
+
+    /// Finalize a PTB. If there were errors during the construction of the PTB these are returned
+    /// now. Otherwise, the PTB is finalized and returned.
+    /// If the warn_on_shadowing flag was set, then we will print warnings for any shadowed
+    /// variables that we encountered during the building of the PTB.
+    pub fn finish(
+        mut self,
+        warn_on_shadowing: bool,
+    ) -> Result<Tx::ProgrammableTransaction, Vec<PTBError>> {
+        if warn_on_shadowing {
+            for (ident, commands) in self.identifiers.iter() {
+                if commands.len() == 1 {
+                    continue;
+                }
+
+                for (i, command_loc) in commands.iter().enumerate() {
+                    if i == 0 {
+                        self.errors.push(PTBError {
+                            message: format!("Variable '{}' first declared here", ident),
+                            span: *command_loc,
+                            help: None,
+                        });
+                    } else {
+                        self.errors.push(PTBError {
+                            message: format!(
+                                "Variable '{}' used again here (shadowed) for the {} time.",
+                                ident, to_ordinal_contraction(i + 1)
+                            ),
+                            span: *command_loc,
+                            help: Some("You can either rename this variable, or do not \
+                                       pass the `warn-shadows` flag to ignore these types of errors.".to_string()),
+                        });
+                    }
+                }
+            }
+        }
+
+        if !self.errors.is_empty() {
+            return Err(self.errors);
+        }
+
+        let ptb = self.ptb.finish();
+        Ok(ptb)
+    }
+
+    pub async fn build(
+        mut self,
+        program: Program,
+    ) -> Result<Tx::ProgrammableTransaction, Vec<PTBError>> {
+        for command in program.commands.into_iter() {
+            self.handle_command(command).await;
+        }
+        self.finish(program.warn_shadows_set)
+    }
+
+    /// Add a single PTB command to the PTB that we are building up.
+    /// Errors are added to the `errors` field of the PTBBuilder.
+    async fn handle_command(&mut self, sp!(span, command): Spanned<ParsedPTBCommand>) {
+        if let Err(e) = self.handle_command_(span, command).await {
+            self.errors.push(e);
+        }
+    }
+
+    // ===========================================================================
+    // Declaring and handling identifiers and variables
+    // ===========================================================================
+
+    /// Declare an identifier. This is used to support shadowing warnings.
```

Comment:
I don't really get what this does. It simply keeps track of identifiers and their associated span? I assume this is used mostly for warning of shadows, and pointing to the right identifier?

Reply:
Yep! Exactly that!